Source/Particles/Resampling/LevelingThinning.cpp

```diff
@@ -0,0 +1,94 @@
+/* Copyright 2019-2020 Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "LevelingThinning.H"
+#include "Particles/Collision/CollisionType.H"
+
+#include <AMReX_Particles.H>
+
+LevelingThinning::LevelingThinning ()
+{
+    using namespace amrex::literals;
+
+    amrex::ParmParse pp("resampling_algorithm");
+    pp.query("target_ratio", m_target_ratio);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE( m_target_ratio > 0._rt,
+                                    "Resampling target ratio should be strictly greater than 0");
+    if (m_target_ratio <= 1._rt)
+    {
+        amrex::Warning("WARNING: target ratio for leveling thinning is smaller or equal to one."
+                       " It is possible that no particle will be removed during resampling");
+    }
+}
+
+void LevelingThinning::operator() (WarpXParIter& pti, const int lev,
+                                   WarpXParticleContainer * const pc) const
+{
+    using namespace amrex::literals;
+
+    auto& ptile = pc->ParticlesAt(lev, pti);
+    auto& soa = ptile.GetStructOfArrays();
+    amrex::ParticleReal * const AMREX_RESTRICT w = soa.GetRealData(PIdx::w).data();
+    WarpXParticleContainer::ParticleType * const AMREX_RESTRICT
+                                 particle_ptr = ptile.GetArrayOfStructs()().data();
+
+    // Using this function means that we must loop over the cells in the ParallelFor. In the case
+    // of the leveling thinning algorithm, it would have possibly been more natural and more
+    // efficient to directly loop over the particles. Nevertheless, this structure with a loop over
+    // the cells is more general and can be readily used to implement almost any other resampling
+    // algorithm.
+    auto bins = findParticles::findParticlesInEachCell(lev, pti, ptile);
+
+    const int n_cells = bins.numBins();
+    const auto indices = bins.permutationPtr();
+    const auto cell_offsets = bins.offsetsPtr();
+
+    const amrex::Real target_ratio = m_target_ratio;
+
+        // Loop over cells
+    amrex::ParallelFor( n_cells,
+        [=] AMREX_GPU_DEVICE (int i_cell) noexcept
+        {
+            // The particles that are in the cell `i_cell` are
+            // given by the `indices[cell_start:cell_stop]`
+            const auto cell_start = cell_offsets[i_cell];
+            const auto cell_stop  = cell_offsets[i_cell+1];
+            const int cell_numparts = cell_stop - cell_start;
+
+            amrex::Real average_weight = 0._rt;
+
+            // First loop over cell particles to compute average particle weight in the cell
+            for (int i = cell_start; i < cell_stop; ++i)
+            {
+                average_weight += w[indices[i]];
+            }
+            average_weight /= cell_numparts;
```

Comment:
I think this might lead to a division by zero for vacuum cells.

We could do:
```suggestion
            if (cell_numparts > 0)
                average_weight /= cell_numparts;
```

or return early (see above).

Reply:
Thanks for noting this. I've committed your return early solution.